ark/type/keywords/string/base64.ts

```diff
@@ -1,24 +1,125 @@
+import { rootSchema, type TraversalContext } from "@ark/schema"
 import type { Module, Submodule } from "../../module.ts"
-import type { Predicate, of } from "../inference.ts"
+import type { Predicate, To, of } from "../inference.ts"
 import { arkModule } from "../utils.ts"
 import { regexStringNode } from "./utils.ts"
 
+// Using JS-only solution to parse Base64 as `Buffer.from` isn't available in browsers and `btoa` is
+// notoriously slow.
+//
+// Code adapted from base64-js: https://github.com/feross/base64-js/blob/master/index.js
+
+const lookup =
+	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".split("")
+const revLookup = lookup.reduce(
+	(obj, char, i) => {
+		obj[char.charCodeAt(0)] = i
+		return obj
+	},
+	{} as Record<number, number>
+)
+
+const getLens = (b64: string) => {
+	const len = b64.length
+
+	if (len % 4 > 0)
+		throw new SyntaxError("Invalid string. Length must be a multiple of 4")
+
+	// Trim off extra bytes after placeholder bytes are found
+	// See: https://github.com/beatgammit/base64-js/issues/42
+	let validLen = b64.indexOf("=")
+	if (validLen === -1) validLen = len
+
+	const placeHoldersLen = validLen === len ? 0 : 4 - (validLen % 4)
+
+	return [validLen, placeHoldersLen]
+}
+
+const byteLength = (validLen: number, placeHoldersLen: number) =>
+	((validLen + placeHoldersLen) * 3) / 4 - placeHoldersLen
+
+const parseB64 = (b64: string) => {
```

Comment:
Hmm. I'm tempted to just say let's use the builtin (btoa) even if it's much slower because people complaining about the perf could just write their own version of the parse that uses one of these libraries, but this is a lot to maintain and add to the bundle size for a perf improvement (albeit significant) on a relatively niche feature.

Another option for this kind of type would be to create a new package like `@ark/extras` where they could be imported directly if users need them. If they're builtin to the default scope, there's no way to tree shake it so the cost is high.

I do appreciate the work that went into adapting an optimized version of this, so would rather see it available through a package like that than to go to waste.

Reply:
I'd argue that the feature wouldn't be that niche as uploading binary data in the form of a base64 string is fairly common, and I'd hate for people to use `string.base64.parse` on the promise that it automatically validates and parses the string into a buffer-like object only to discover it's using an approach that's 10,000% slower than an available alternative (and given the nature of how this is likely to be used, that slowdown could quickly add up). In my honest opinion, from a user experience perspective, I think it would be better for the feature to not exist at all than exist in such a disadvantageous state.

That said, the maintenance and bundle size arguments are valid, so maybe an `@ark/extras` package is warranted if this PR is to go further.

Reply:
Yeah, especially until we can support precompilation as a build step for people who care about bundle size, we have to be pretty conservative about this sort of thing.

If you want to migrate this with the changes mentioned to `@ark/extras`, I would merge that (although the name might change before publishing).